pkg/api/write.go

```diff
@@ -158,32 +165,60 @@ func decodeSnappy(w http.ResponseWriter, r *http.Request) bool {
 	if detectSnappyStreamFormat(buf) {
 		r.Body = &readCloser{
 			reader: snappy.NewReader(mr),
-			closer: r.Body,
+			closer: originalBody,
 		}
 		return true
 	}
 
-	compressed, err := ioutil.ReadAll(mr)
+	compressed := compressedBufPool.Get().(*bytes.Buffer)
+	compressed.Reset()
+	defer compressedBufPool.Put(compressed)
+
+	_, err = compressed.ReadFrom(mr)
 	if err != nil {
 		invalidRequestError(w, "request body read error", err.Error(), metrics)
 		return false
 	}
 
+	b := decodedBufPool.Get().(*bytes.Buffer)
+	b.Reset()
```

Comment:
I disagree. And the example in sync.Pool disagrees. It's always safer to clear objects after Get. See this SO answer: https://stackoverflow.com/a/62677883/8483666

```Second, per the examples, the best practice is to reset the pooled object after Get, not before Put. This is safest because it means Get is not making any potentially unsafe assumptions about the values it receives from the pool. You could also do both if items tend to stay in the pool for a long time and are quite large (there's a potential here for very high memory usage from slices of maps, so this will depend on your use case).``` 

For the Garbage collector it doesn't matter which way you do it because you are holding on to the same memory.

Reply:
I'll also add that if your reset when you Put that logic may also be needed in the New function of the pool. This is a bug we saw in your exemplar PR. If you do it after Get then the new and pooled case are guaranteed to be treated the same. That makes reset after Get safer.

Reply:
The link you mentioned is good if we are getting something from pool, whose default value, as set by new(), is not the default value expected.

`Get is not making any potentially unsafe assumptions about the values it receives from the pool`

Your line itself says that. The last part of that qoute is saying what I had said, keeping bigger things in a pool will lead to mem increase, as the pool keeps holding larger objects unless it reaches an equilibrium. I don't understand why you disagree.

Reply:
> This is a bug we saw in your exemplar PR.

Can you link the part here? I don't recall any place  in my exemplar PR that was having non default Get() values, otherwise tests would have failed, but still.

Reply:
The thing I said logically makes sense. Please note, the default values should satisfy a usecase, otherwise it's a bug. Can you share that sync.Pool example? Would be interesting to see where Get() returned alloc should be reset/optimized in default value usecase.